# AI venue type investment model
## [Carlos Morlan](https://www.linkedin.com/in/carlos-morlan-96343a15/)
### Published date: July 27<sup>th</sup>, 2019

[![Battle of neighborhoods](https://www.garybarker.co.uk/files/uk-city-life-cartoon-illustration.jpg)](https://www.garybarker.co.uk)

> _The ultimate purpose of analytics is to communicate findings to the concerned who might use these insights to formulate policy or strategy.
> The_ _<b>data scientist</b> should then use the insights to build the narrative to communicate the findings._ <font size="1">Murtaza Haider, (2015) [Getting Started with Data Science: Making Sense of Data with Analytics](https://books.google.com.mx/books?id=b4YxCwAAQBAJ&printsec=frontcover). IBM Press.</small></font>

# Table of contents<a name="TaOCo"></a>

  - [Executive Summary](#ExSum)
  - [Introduction](#Intro)
  - [Data acquisition and cleaning](#DAcaC)
  - [Methodology](#Metho)
  - [Detailed results & discussion](#DetRe)
  - [Conclusions](#Concl)
  - [References](#Refer)

### <a name="ExSum"></a>Executive Summary

I can tell you by experience that as a full time employee a day will come where you start thinking on becoming your own boss by opening a business. You have enough money, or at least you think that you have it. You are almost an expert on how to manage the business type you have in mind, or at least you believe so. So why not? Well, as soon as you start going into the details of your _original_ plan, you realize that you may not be so expert to open the business you choose, but smart enough to learn about it. Or maybe your family and/or friends give you a better option to pick your *future* business category. Now you have more than one option on which business type will help you be the greatest businessman for whom you were born to be.

Now the second question to answer is on which part of my city should I invest my money to open my new business? Sometimes opening a business in a neighborhood where there are no similar businesses nearby can be a good idea. On the other hand, it is best to open a business near other businesses whose customers can be mine too. For example, opening a pet food store not far away from a veterinary clinic can ensure the success of my business.

As you can see, before going on more details on our business plan, finding a list, book, website or tool where I can identify what types of business exist on every neighborhood in my city will make my life easier to start my new business, right? The intend of this document is to show how Data Science can be implemented to build this tool regardless the city where you are from in the whole world. In particular, I will focus on mine, *Mexico City*.

<font size="1">[Go to Table of contents](#TaOCo)</font>

### <a name="Intro"></a>Introduction

I live in Mexico City, one of the biggest and most populated cities in the world. One of its citizens main concerns is that the country's economy is volatile, you can feel it in the air. A good proxy for the overall stability of a country is the consistency of its economic growth. From my personal point of view, getting more investments is a good way to improve its economic growth. The investments, that can be done by the government or by private companies, should be well planned based on the different communities needs through all the main country's cities.

This capstone project will try to show how Mexico City can attract new invests for Mexico's economy improvement. The Government and new investors should know what are the popular places where the citizens have fun, get dinner or bought supplies. With such information, either of both can make best decisions about the type of business they can open and how well the people will take the new venue. Moreover, if there are popular places with special attributes like a Medical Center, they can start opening required business types near such places like Laboratories, Pharmacies or even a Hotel so the people from outside town can stay there while their patients are receiving treatment. As you can see, small or big investors can use this valuable information to take the path of a successful opportunity and the city communities will also get more and better services: <b>Is a win-win situation</b>.

<font size="1">[Go to Table of contents](#TaOCo)</font>

### <a name="DAcaC">Data acquisition and cleaning</a>

#### <a name="DaSou">Data sources</a>
Some ingredients are mandatory before we can make this happen. A couple of data sources have been identified for this project:

- [Major city spots identified by zone or neighborhood](#MCPbZ). 
- [Popular venues identification based on social networks (www.foursquare.com)](#PviFS).

##### <a name="MCPbZ">Major City Spots</a>
To get this information I will be using a public dataset from www.geonames.org site downloaded locally in case the file is removed or changed in the future, check the [References](#Refer) section for more details. The data format is tab-delimited text in utf8 encoding, with the following fields:

* country code      : iso country code, 2 characters
* postal code       : varchar(20)
* place name        : varchar(180)
* admin name1       : 1. order subdivision (state) varchar(100)
* admin code1       : 1. order subdivision (state) varchar(20)
* admin name2       : 2. order subdivision (county/province) varchar(100)
* admin code2       : 2. order subdivision (county/province) varchar(20)
* admin name3       : 3. order subdivision (community) varchar(100)
* admin code3       : 3. order subdivision (community) varchar(20)
* latitude          : estimated latitude (wgs84)
* longitude         : estimated longitude (wgs84)
* accuracy          : accuracy of lat/lng from 1=estimated, 4=geonameid, 6=centroid of addresses or shape

_FYI, this file doesn't have column headers._

After looking into the data, I chose a better name for each column based on its content and identified which columns can be dropped because are not required for this analysis. The final mapping is shown below:

|Data Source Field|New Column Name|
|:----------|----------|
|postal code|PostalCode|
|place name|PlaceName|
|admin code1|StateCode|
|latitude|Latitude|
|longitude|Longitude|
|accuracy|Accuracy|

Reading the file content using pandas package looks like this:

In [1]:
import pandas as pd # To manage dataframes
import numpy as np # To manage arrays

# Read source, the file is tab delimited and the postal code column (#2) should be treated as string
postal_codes_tmp = pd.read_csv('MX.txt', sep='\t', header=None, dtype={1:str})

# Assign column headers because the file doesn't have it
postal_codes_tmp.columns = ['CountryCode', 'PostalCode', 'PlaceName', 'State', 'StateCode', 'TownHall', 'TownHallCode', 'AdminName3', 'AdminCode3', 'Latitude', 'Longitude', 'Accuracy']
# print(postal_codes_tmp.dtypes)

postal_codes_tmp.head()

CountryCode PostalCode          PlaceName           State  StateCode  \
0          MX      20000        Zona Centro  Aguascalientes          1   
1          MX      20010   Olivares Santana  Aguascalientes          1   
2          MX      20010  Ramon Romo Franco  Aguascalientes          1   
3          MX      20010         Las Brisas  Aguascalientes          1   
4          MX      20010       San Cayetano  Aguascalientes          1   

         TownHall  TownHallCode      AdminName3  AdminCode3  Latitude  \
0  Aguascalientes             1  Aguascalientes         1.0   21.8734   
1  Aguascalientes             1  Aguascalientes         1.0   21.9644   
2  Aguascalientes             1  Aguascalientes         1.0   21.9644   
3  Aguascalientes             1  Aguascalientes         1.0   21.9644   
4  Aguascalientes             1  Aguascalientes         1.0   21.9644   

   Longitude  Accuracy  
0  -102.2806         1  
1  -102.3192         1  
2  -102.3192         1  
3  -102.3192         1  
4  -102.3192         1

<font size="1">[Go to Data sources](#DaSou)</font>

<font size="1">[Go to Table of contents](#TaOCo)</font>

##### <a name="DatCl">Data cleaning</a>

Regardless that the data for every mexican neighborhood is available, I will apply a couple of filters to the data source because of 2 reasons:

- The analysis that will be done is for Mexico City only (i.e. `admin code1 == 9`).
- The number of Foursquare API free calls are limited by day and won't suffice to process all the Mexico cities (i.e. 50 unique coordinates will be used, latitude/longitude values).

Whenever this process is productionable the filters can be easily removed to analyze the whole country if needed.

The Major City spots data has several place names for the same latitude and longitude because some of the coordinates are estimated (check Accuracy column definition). Because of this reason, a unique combination for latitude and longitude is identified and a list of place names delimited by comma is created as a new column. I'm leaving a record count column to debug the process, but is not needed for the analysis.

In [3]:
# Add leading zeros to the Postal code
# postal_codes_tmp['PostalCode'] = postal_codes_tmp['PostalCode'].apply(lambda x: x.zfill(5))

# Add filters to process a single state town hall for testing purposes
state_filter = postal_codes_tmp['StateCode']==9
#townhall_filter = postal_codes_tmp['TownHallCode']==3
#filtered_postal_codes = postal_codes_tmp[state_filter & townhall_filter]
filtered_postal_codes = postal_codes_tmp[state_filter]
#filtered_postal_codes.head()

# Remove unused columns
filtered_postal_codes.drop(columns=['CountryCode', 'PostalCode', 'State', 'StateCode', 'TownHall', 'TownHallCode', 'AdminName3', 'AdminCode3'], inplace=True)

# Leave only one latitude-longitude by PostalCode (mean will be used)
#unique_coordinates = filtered_postal_codes.groupby('PostalCode').agg({'PlaceName': [(', '.join)], 'Latitude': 'mean', 'Longitude': 'mean', 'Accuracy': 'count'}).reset_index()

# Leave unique latitude-longitude combination; otherwise, the map will show only one point for multiple places
unique_coordinates = filtered_postal_codes.groupby(['Latitude','Longitude']).agg({'PlaceName': [(', '.join)], 'Accuracy': 'count'}).reset_index()

# Rename column headers
unique_coordinates.columns = ['Latitude', 'Longitude', 'PlaceName', 'RecordCount']

The final dataframe, `major_city_spots`, needed for this model has the following columns:

- PlaceName
- Latitude
- Longitude
- RecordCount

In [4]:
# Rearrange columns
cols = ['PlaceName', 'Latitude', 'Longitude', 'RecordCount']
unique_coordinates = unique_coordinates[cols]

# Kmoso filter to be removed
major_city_spots = unique_coordinates.head(50)

# Verify data frame consistency
print('{} rows in dataframe'.format(major_city_spots.shape[0]))
total_places = len(major_city_spots['PlaceName'].unique())
print('{} unique places'.format(total_places))
major_city_spots.head()

50 rows in dataframe
50 unique places


PlaceName  Latitude  Longitude  \
0                                   Parres El Guarda   19.1361   -99.1738   
1  La Concepción, San Mateo, Los Ángeles, Emilian...   19.1395   -99.0511   
2                                         San Marcos   19.1694   -99.0257   
3                             San Lorenzo Tlacoyucan   19.1761   -99.0322   
4                      San Juan, San Juan Tepenahuac   19.1877   -98.9945   

   RecordCount  
0            1  
1           10  
2            1  
3            1  
4            2

Below you can see a Mexico City map where each dataframe place has a marker.

In [7]:
#!conda install -c conda-forge geopy --yes # uncomment this line if you haven't installed geopy package
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

# Verify if below command should be executed in the current environment
# !conda update -n base -c defaults conda

#!conda install -c conda-forge folium=0.5.0 --yes # uncomment this line if you haven't completed the Foursquare API lab
import folium # map rendering library

import requests # library to handle requests

# import k-means from clustering stage
from sklearn.cluster import KMeans

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

address = 'Mexico City, Mexico'

geolocator = Nominatim(user_agent="city_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geographical coordinates of {} are {}, {}.'.format(address, latitude, longitude))

# create map of the chosen city using latitude and longitude values
map_city = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, placename in zip(major_city_spots['Latitude'], major_city_spots['Longitude'], major_city_spots['PlaceName']):
    label = '{}'.format(placename)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_city)  
    
map_city

The geograpical coordinates of Mexico City, Mexico are 19.4326009, -99.1333416.


<font size="1">[Go to Data acquisition and cleaning](#DAcaC)</font>

<font size="1">[Go to Table of contents](#TaOCo)</font>

##### <a name="PviFS">Popular Venues Identification</a>
A developer account was created in Foursquare to get access to the available endpoints to get the popular venues from a particular location and radius, check the [References](#Refer) section for more details.

_FYI For this project a 500 meter radius will be used unless something different is noted._

One single Foursquare endpoint will be used to get the places nearby a specific location defined by its latitude and longitude. The search API call will return a JSON object that should be read to identify the venue's attributes: name, latitude, longitude and category. The category is what will identify the business type of the venue. Take into account that the identified venues are those which the Foursquare users have been checked-in a visit.

In [8]:
# Define Foursquare Credentials and Version
CLIENT_ID = '****************************'
CLIENT_SECRET = '****************************'
VERSION = '****************************'
LIMIT = 10

A couple of Python functions were created, `searchNearbyVenues()` to read the JSON object returned by the Foursquare API and `return_most_common_venues()` to sort the most common venues nearby a specific place order by its frequency (check-in records). In particular, the first function manage the exception whenever for the specific place there are no venues around.

In [9]:
# Below function is not used for this proyect
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print('Getting data for ' + name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PlaceName', 
                  'PlaceName Latitude', 
                  'PlaceName Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

def searchNearbyVenues(names, latitudes, longitudes, radius=500, intent='checkin'):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print('Getting data for ' + name)

        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/search?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&intent={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT,
            intent)

        # make the GET request
        results = requests.get(url).json()["response"]['venues']

        try:
            # return only relevant information for each nearby venue
            venues_list.append([(
                name, 
                lat, 
                lng, 
                v['name'], 
                v['location']['lat'], 
                v['location']['lng'],  
                v['categories'][0]['name']) for v in results])
        except Exception as e:
            print('Exception when calling ' + url + ' with (lat,lon) = (' + str(lat) + ',' + str(lng) + ')')
            print(str(e))
            pass
    
    try:
        nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
        nearby_venues.columns = ['PlaceName', 
                      'PlaceName Latitude', 
                      'PlaceName Longitude', 
                      'Venue', 
                      'Venue Latitude', 
                      'Venue Longitude', 
                      'Venue Category']
    except Exception as e:
        print('Exception when creating dataframe')
        print(str(e))
        nearby_venues = pd.DataFrame()
        pass
    
    return(nearby_venues)

This is the API call output for the 6th record of a sample location, take into account that the client_id, client_secret and API version are mandatory parameters (hidden in the image for security purposes).

In [10]:
# Sample API call
sample_url = 'https://api.foursquare.com/v2/venues/search?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}&intent={}'.format(
      CLIENT_ID, 
      CLIENT_SECRET, 
      VERSION, 
      19.353023,
      -99.155828, 
      500, 
      LIMIT,
      'checkinintent')

# make the GET request
sample_results = requests.get(sample_url).json()["response"]['venues']
sample_results[5]

{'id': '522e010f11d2740eb812b26f',
 'name': 'Laboratorio Biologia A-23A',
 'location': {'lat': 19.353087092842518,
  'lng': -99.15553406100847,
  'labeledLatLngs': [{'label': 'display',
    'lat': 19.353087092842518,
    'lng': -99.15553406100847}],
  'distance': 31,
  'cc': 'MX',
  'country': 'México',
  'formattedAddress': ['México']},
 'categories': [{'id': '4bf58dd8d48988d1a5941735',
   'name': 'College Lab',
   'pluralName': 'College Labs',
   'shortName': 'Lab',
   'icon': {'prefix': 'https://ss3.4sqi.net/img/categories_v2/education/lab_',
    'suffix': '.png'},
   'primary': True}],
 'referralId': 'v-1564269326',
 'hasPerk': False}

### <a name="Metho"></a>Methodology

With the acquired Data Science knowledge from all the courses included in the IBM Data Science Professional Coursera track an AI venue type investment model will be created. What information will be shown by this model? First of all, a map that will display 7 different clusters that group neighborhoods with similar venue types. As learned through all the past weeks, data visualization is always more advantageous  to communicate ideas. Secondly, for each cluster a list of most common venue types will be shown ordered by it's frequency. With these lists, the Government or private investors can decide where to open a new business (on which neighborhood) and of what type (based on the venue's category from the social network used). They can also notice if there is any type of missing business required on the specific neighborhoods.

#### Data Science Tools & Algorithms
Python Notebook was created in Skills Network Labs framework for this project, check the [References](#Refer) section for more details.

k-means clustering method will be used to group the data identifying the 10 more common venues in 500 meters radius for each identified zone/neighborhood. After identifying all the venues, the data for each cluster will be grouped by category's venue to show the most frequent categories.

_FYI A matrix of 10 x n will be created by the algorithm (number of the most common venues times the number of neighborhoods) for each cluster_

<font size="1">[Go to Table of contents](#TaOCo)</font>

### <a name="DetRe">Detailed results & discussion</a>
Now that the Major Spots are identified and the functions are ready to be used let's get all the city venues based on the Major City Spots dataframe.

In [11]:
# Defining dataframe number of columns
city_venues = pd.DataFrame(pd.np.empty((0, 7)))

# Get nearby venues data using explore end point
#city_venues = getNearbyVenues(names=major_city_spots['PlaceName'],
#                                   latitudes=major_city_spots['Latitude'],
#                                   longitudes=major_city_spots['Longitude']
#                                  )

# Get nearby venues data using search end point
city_venues = searchNearbyVenues(names=major_city_spots['PlaceName'],
                                   latitudes=major_city_spots['Latitude'],
                                   longitudes=major_city_spots['Longitude']
                                  )

Getting data for Parres El Guarda
Exception when calling https://api.foursquare.com/v2/venues/search?&client_id=020DHIJQ5OJ4YZ12HXXY4O0D33CXV4OT0QXK25QO3Y03IK1I&client_secret=P3SIW32METMPEVCC1WEZ3DXWQEFVGA2YZC5ELTDWT2FSYVW4&v=20180605&ll=19.1361,-99.1738&radius=500&limit=10&intent=checkin with (lat,lon) = (19.1361,-99.1738)
list index out of range
Getting data for La Concepción, San Mateo, Los Ángeles, Emiliano Zapata, La Conchita, Tula, San Miguel, Centro, Chalmita, San Miguel
Getting data for San Marcos
Exception when calling https://api.foursquare.com/v2/venues/search?&client_id=020DHIJQ5OJ4YZ12HXXY4O0D33CXV4OT0QXK25QO3Y03IK1I&client_secret=P3SIW32METMPEVCC1WEZ3DXWQEFVGA2YZC5ELTDWT2FSYVW4&v=20180605&ll=19.1694,-99.0257&radius=500&limit=10&intent=checkin with (lat,lon) = (19.1694,-99.0257)
list index out of range
Getting data for San Lorenzo Tlacoyucan
Exception when calling https://api.foursquare.com/v2/venues/search?&client_id=020DHIJQ5OJ4YZ12HXXY4O0D33CXV4OT0QXK25QO3Y03IK1I&client

In [12]:
print(str(city_venues.shape[0]) + ' venues with ' + str(city_venues.shape[1]) + ' columns')
print('There are {} uniques categories.'.format(len(city_venues['Venue Category'].unique())))

city_venues.head(12)

330 venues with 7 columns
There are 117 uniques categories.


PlaceName  PlaceName Latitude  \
0   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
1   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
2   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
3   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
4   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
5   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
6   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
7   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
8   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
9   La Concepción, San Mateo, Los Ángeles, Emilian...             19.1395   
10                      San Juan, San Juan Tepenahuac             19.1877   
11                      San Juan, San Juan Tepenahuac             19.1877   

    PlaceName Longitude                          Venue  Venue Latitude  \
0              -99.0511  Deportivo San Pablo Oztotepec       19.184272   
1              -99.0511                     LURIASROCK       19.176086   
2              -99.0511            Advanced Deisel LLC       19.138053   
3              -99.0511                   CICS UMA IPN       19.174541   
4              -99.0511                pinturas "zaer"       19.185582   
5              -99.0511                        lidecue       19.122747   
6              -99.0511     Gasera San Pablo Oztotepec       19.183723   
7              -99.0511   Estadio De Fútbol Milpa Alta       19.192259   
8              -99.0511          Iglesia Barrio La Luz       19.188190   
9              -99.0511             Mirador Milpa Alta       19.186763   
10             -98.9945     Iglesia Del Señor San Juan       19.187234   
11             -98.9945          Primaria Sierra Leona       19.192236   

    Venue Longitude        Venue Category  
0        -99.069881          Soccer Field  
1        -99.031683                 Plaza  
2        -99.058506       Automotive Shop  
3        -99.062473        Medical School  
4        -99.074558  Outdoor Supply Store  
5        -98.998848          Soccer Field  
6        -99.066370           Gas Station  
7        -99.023480        Soccer Stadium  
8        -99.027650                Church  
9        -99.020860            Food Court  
10       -98.994049                Church  
11       -99.000657        Student Center

Based on this dataframe, `city_venues_grouped` dataframe was generated to identify the venues categories for each major spot. You can think about it as a matrix of n rows and m columns where n is the number of distinct major spots and m is the number of categories (in this case a 33 x 118 matrix).

In [13]:
city_venues_categories = pd.get_dummies(city_venues[['Venue Category']], prefix="", prefix_sep="")

# add place name column back to dataframe
city_venues_categories['PlaceName'] = city_venues['PlaceName'] 

# move place name column to the first column
fixed_columns = [city_venues_categories.columns[-1]] + list(city_venues_categories.columns[:-1])
city_venues_categories = city_venues_categories[fixed_columns]

# city_venues_categories.head()
print('{} places and {} categories before grouping by Place Name'.format(city_venues_categories.shape[0], city_venues_categories.shape[1]))

city_venues_grouped = city_venues_categories.groupby('PlaceName').mean().reset_index()

print('{} places and {} categories after grouping by Place Name'.format(city_venues_grouped.shape[0], city_venues_grouped.shape[1]))
city_venues_grouped.tail()

330 places and 118 categories before grouping by Place Name
33 places and 118 categories after grouping by Place Name


PlaceName  African Restaurant  \
28                                         Tenantitla                 0.0   
29                                    Tepantitlamilco                 0.0   
30                Tlaltepetla, Santa Cruz Chavarrieta                 0.0   
31  Tlaxopan  2a Sección, Tlaxopan 1a Sección, Tla...                 0.0   
32                                   Villa Xochimilco                 0.1   

    Argentinian Restaurant  Art Gallery  Assisted Living  Athletics & Sports  \
28                     0.0          0.0              0.0                 0.0   
29                     0.0          0.0              0.0                 0.1   
30                     0.0          0.0              0.1                 0.0   
31                     0.0          0.0              0.0                 0.0   
32                     0.1          0.0              0.0                 0.0   

    Auto Dealership  Auto Garage  Automotive Shop  BBQ Joint  ...  Stables  \
28              0.0          0.0              0.0        0.0  ...      0.0   
29              0.0          0.0              0.0        0.0  ...      0.0   
30              0.0          0.0              0.0        0.0  ...      0.1   
31              0.0          0.0              0.0        0.0  ...      0.0   
32              0.0          0.0              0.0        0.0  ...      0.0   

    Stationery Store  Street Fair  Strip Club  Student Center  Taco Place  \
28               0.0          0.0         0.0             0.0         0.0   
29               0.0          0.0         0.0             0.1         0.0   
30               0.0          0.0         0.0             0.0         0.0   
31               0.0          0.0         0.0             0.0         0.0   
32               0.0          0.0         0.0             0.0         0.0   

    Temple  Toll Booth  Trail  University  
28     0.0         0.0    0.0         0.0  
29     0.0         0.0    0.0         0.0  
30     0.0         0.0    0.0         0.1  
31     0.0         0.1    0.0         0.0  
32     0.0         0.0    0.0         0.0  

[5 rows x 118 columns]

Let's display the data in a different way. For a couple of Major City Spots the venue category frequency will be shown.

In [14]:
num_top_venues = 10

for hood in city_venues_grouped['PlaceName']:
    temp = city_venues_grouped[city_venues_grouped['PlaceName'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    if (hood == 'San José') | (hood == 'Ocotitla'):
        print("----"+hood+"----")
        print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
        print('\n')

----Ocotitla----
                venue  freq
0  Mexican Restaurant   0.2
1                Bank   0.2
2         Pizza Place   0.1
3          Taco Place   0.1
4               Plaza   0.1
5    Basketball Court   0.1
6      Soccer Stadium   0.1
7     Auto Dealership   0.1
8         High School   0.0
9      Medical School   0.0


----San José----
                    venue  freq
0      Mexican Restaurant   0.5
1          Clothing Store   0.1
2  Argentinian Restaurant   0.1
3                  Garden   0.1
4            Soccer Field   0.1
5                    Pool   0.1
6           Movie Theater   0.0
7          Medical School   0.0
8      Miscellaneous Shop   0.0
9                Mountain   0.0




The next step is to create the final dataframe `city_venues_sorted` to visualize for each Major Spot the 10th most common venues. Later I will add the cluster associated to each row.

In [15]:
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns_newdf = ['PlaceName']
for ind in np.arange(num_top_venues):
    try:
        columns_newdf.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns_newdf.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
city_venues_sorted = pd.DataFrame(columns=columns_newdf)
city_venues_sorted['PlaceName'] = city_venues_grouped['PlaceName']

for ind in np.arange(city_venues_grouped.shape[0]):
    city_venues_sorted.iloc[ind, 1:] = return_most_common_venues(city_venues_grouped.iloc[ind, :], num_top_venues)

city_venues_sorted.head()

PlaceName  \
0                      Cabeza de Juárez 5 o Frente 5   
1                                            Cantera   
2                                          Cruztitla   
3  Culhuacán CTM Sección IX-A, Culhuacán CTM Secc...   
4  Degollado, La Magueyera, Prados, Arboledas Zaf...   

                      1st Most Common Venue 2nd Most Common Venue  \
0                                 BBQ Joint           High School   
1                        Mexican Restaurant     French Restaurant   
2                                 BBQ Joint                 Field   
3                        Mexican Restaurant        Clothing Store   
4  Residential Building (Apartment / Condo)         Movie Theater   

  3rd Most Common Venue 4th Most Common Venue         5th Most Common Venue  \
0  Other Great Outdoors           Art Gallery                 Movie Theater   
1        Scenic Lookout     Convenience Store                      Mountain   
2                Garden                 Diner                           Bar   
3          Soccer Field           Street Fair                       Brewery   
4                Temple    College Rec Center  General College & University   

    6th Most Common Venue  7th Most Common Venue 8th Most Common Venue  \
0          Student Center                  Field   Housing Development   
1               Rest Area        College Library    College Rec Center   
2             Event Space  General Entertainment            Taco Place   
3  Argentinian Restaurant      Convenience Store    Country Dance Club   
4   Outdoors & Recreation                 Church   Housing Development   

     9th Most Common Venue 10th Most Common Venue  
0                Rest Area               Pharmacy  
1  Comfort Food Restaurant     Country Dance Club  
2               Food Truck    Distribution Center  
3  Comfort Food Restaurant        Cultural Center  
4                Speakeasy                    Spa

<font size="1">[Go to Detailed results & discussion](#DetRe)</font>

<font size="1">[Go to Table of contents](#TaOCo)</font>

#### k-means Clustering

Everything is now set to identify 7 clusters to group the venues and display them in a map. The first step is to drop the place name from the data because otherwise it will only cause noise. Now let's use the existing city_venues_sorted dataframe and add a new column to identify on which cluster every major spot will be grouped. Finally, to make the dataframe more readable a join with the city_major_spots was done to show the latitude and longitude as well.

In [18]:
# set number of clusters
kclusters = 7

city_venues_grouped_clustering = city_venues_grouped.drop('PlaceName', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(city_venues_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

# add clustering labels
city_venues_sorted.insert(0, 'ClusterL', kmeans.labels_)

city_venues_merged = major_city_spots

# merge city_venues_grouped with places to add latitude/longitude for each place
city_venues_merged = city_venues_merged.join(city_venues_sorted.set_index('PlaceName'), on='PlaceName')
city_venues_merged.fillna(99, inplace=True) # This will identify centroids which "die" while iterating and can be safely excluded
# city_venues_merged.dtypes

# store places not classified for future analysis in a different dataframe
city_venues_merged['ClusterL'] = city_venues_merged['ClusterL'].astype(int)
city_venues_wo_cluster = city_venues_merged[city_venues_merged['ClusterL']==99]
city_venues_merged = city_venues_merged[city_venues_merged['ClusterL']<99]

# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(city_venues_merged['Latitude'], city_venues_merged['Longitude'], city_venues_merged['PlaceName'], city_venues_merged['ClusterL']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

An important explanation must be added to this report. In the k-means clustering iteration process some of the centroids can _die_, meaning that a particular Major Spot can't be grouped to any other because there are not enough information to do so. This scenario happened in this project and those places were removed without impacting the model, check the [References](#Refer) section for more details. A dataframe  `city_venues_wo_cluster` was created to make a separate analysis about those places in a different project.

In [19]:
city_venues_wo_cluster.head()

PlaceName  Latitude  Longitude  RecordCount  ClusterL  \
0             Parres El Guarda   19.1361   -99.1738            1        99   
2                   San Marcos   19.1694   -99.0257            1        99   
3       San Lorenzo Tlacoyucan   19.1761   -99.0322            1        99   
10              Club Monte Sur   19.1960   -99.0902            1        99   
11  San Francisco Tlalnepantla   19.1974   -99.1223            1        99   

   1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0                     99                    99                    99   
2                     99                    99                    99   
3                     99                    99                    99   
10                    99                    99                    99   
11                    99                    99                    99   

   4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0                     99                    99                    99   
2                     99                    99                    99   
3                     99                    99                    99   
10                    99                    99                    99   
11                    99                    99                    99   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                     99                    99                    99   
2                     99                    99                    99   
3                     99                    99                    99   
10                    99                    99                    99   
11                    99                    99                    99   

   10th Most Common Venue  
0                      99  
2                      99  
3                      99  
10                     99  
11                     99

<font size="1">[Go to Table of contents](#TaOCo)</font>

#### The final output

Now for each cluster the most common venues nearby a particular major spot are identified. The investors actually can use this information to make their decisions, they have 7 different groups of venues and at least 10 most common venues for each one, this give them more than 70 different options to chose from. In particular, if you remember the beginning of the [Detailed results & discussion](#DetRe) section, they will have 330 different venues. Let's see how the first cluster looks like:

In [30]:
cluster0 = city_venues_merged.loc[city_venues_merged['ClusterL'] == 0, city_venues_merged.columns[[0] + list(range(5, city_venues_merged.shape[1]))]]
cluster0.head()

PlaceName 1st Most Common Venue  \
1   La Concepción, San Mateo, Los Ángeles, Emilian...          Soccer Field   
27                                San Nicolás Tetelco       Business Center   
28                   Ojo de Agua, Guadalupe Tlaltenco           Event Space   
32                                         Tenantitla          Burger Joint   
33                                    Tepantitlamilco                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1            Gas Station       Automotive Shop            Food Court   
27              Mountain           Coffee Shop            Taco Place   
28                Church                Market           Beer Garden   
32                   Gym                  Park       Cultural Center   
33                 Field               Brewery           Beer Garden   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1         Medical School                Church                 Plaza   
27    Athletics & Sports                  Park            Public Art   
28    Mexican Restaurant              Cemetery        Sandwich Place   
32            Public Art        Medical Center               Library   
33              Cemetery                 Plaza             Juice Bar   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1         Soccer Stadium  Outdoor Supply Store              BBQ Joint  
27           Beer Garden                 Plaza      Elementary School  
28                 Field           Coffee Shop          Movie Theater  
32           High School     College Classroom    Government Building  
33                  Lake    Athletics & Sports         Student Center

Finally, let's check what is the venue distribution by cluster.

In [25]:
cluster1 = city_venues_merged.loc[city_venues_merged['ClusterL'] == 1, city_venues_merged.columns[[0] + list(range(5, city_venues_merged.shape[1]))]]
cluster2 = city_venues_merged.loc[city_venues_merged['ClusterL'] == 2, city_venues_merged.columns[[0] + list(range(5, city_venues_merged.shape[1]))]]
cluster3 = city_venues_merged.loc[city_venues_merged['ClusterL'] == 3, city_venues_merged.columns[[0] + list(range(5, city_venues_merged.shape[1]))]]
cluster4 = city_venues_merged.loc[city_venues_merged['ClusterL'] == 4, city_venues_merged.columns[[0] + list(range(5, city_venues_merged.shape[1]))]]
cluster5 = city_venues_merged.loc[city_venues_merged['ClusterL'] == 5, city_venues_merged.columns[[0] + list(range(5, city_venues_merged.shape[1]))]]
cluster6 = city_venues_merged.loc[city_venues_merged['ClusterL'] == 6, city_venues_merged.columns[[0] + list(range(5, city_venues_merged.shape[1]))]]

print('             Cluster 0: {} rows --> {}% places included'.format(cluster0.shape[0], round(cluster0.shape[0]/total_places*100,2)))
print('             Cluster 1: {} rows --> {}% places included'.format(cluster1.shape[0], round(cluster1.shape[0]/total_places*100,2)))
print('             Cluster 2: {} rows --> {}% places included'.format(cluster2.shape[0], round(cluster2.shape[0]/total_places*100,2)))
print('             Cluster 3: {} rows --> {}% places included'.format(cluster3.shape[0], round(cluster3.shape[0]/total_places*100,2)))
print('             Cluster 4: {} rows --> {}% places included'.format(cluster4.shape[0], round(cluster4.shape[0]/total_places*100,2)))
print('             Cluster 5: {} rows --> {}% places included'.format(cluster5.shape[0], round(cluster5.shape[0]/total_places*100,2)))
print('             Cluster 6: {} rows --> {}% places included'.format(cluster6.shape[0], round(cluster6.shape[0]/total_places*100,2)))
print('Venues without cluster: {} rows --> {}% places included'.format(city_venues_wo_cluster.shape[0], round(city_venues_wo_cluster.shape[0]/total_places*100,2)))

             Cluster 0: 9 rows --> 18.0% places included
             Cluster 1: 6 rows --> 12.0% places included
             Cluster 2: 5 rows --> 10.0% places included
             Cluster 3: 6 rows --> 12.0% places included
             Cluster 4: 3 rows --> 6.0% places included
             Cluster 5: 2 rows --> 4.0% places included
             Cluster 6: 2 rows --> 4.0% places included
Venues without cluster: 17 rows --> 34.0% places included


An additional way of grouping the data will be done in case the investors want to know the frequency for each venue category regardless in which position of the most common venues appear. E.g. if you check `cluster0` data shown above, you will notice that _Beer Garden_ venue category appears as the fourth most common venue a couple of times and one more time as the eighth most common venue. On the other hand, _Burger Joint_ venue category appears only one time in the whole dataset as the most common venue for `Tenantitla` Major Spot. Some investors may give more weight to their decision to a venue category that seems more successful because appears more than once in the cluster than another that may be one of a kind.

Let's group the results in `cluster0`, count how many times every venue category is present in the most common venues for all the Major Spots and show the new category list order by the sum of the record count. This is the result of the venue categories in `cluster0`:

In [54]:
#pd.set_option('display.max_rows', 90)
mcv1 = cluster0.groupby('1st Most Common Venue').count()
mcv2 = cluster0.groupby('2nd Most Common Venue').count()
mcv3 = cluster0.groupby('3rd Most Common Venue').count()
mcv4 = cluster0.groupby('4th Most Common Venue').count()
mcv5 = cluster0.groupby('5th Most Common Venue').count()
mcv6 = cluster0.groupby('6th Most Common Venue').count()
mcv7 = cluster0.groupby('7th Most Common Venue').count()
mcv8 = cluster0.groupby('8th Most Common Venue').count()
mcv9 = cluster0.groupby('9th Most Common Venue').count()
mcv10 = cluster0.groupby('10th Most Common Venue').count()
tmcv = mcv1.add(mcv2.add(mcv3.add(mcv4.add(mcv5.add(mcv6.add(mcv7.add(mcv8.add(mcv9.add(mcv10, fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0)
print('{} records in cluster 0, 10 top venues = {} entries'.format(str(cluster0.shape[0]),str(int(tmcv['PlaceName'].sum()))))
tmcv['PlaceName'].sort_values(ascending=False)

9 records in cluster 0, 5 top venues = 90 entries


Plaza                           5.0
Church                          4.0
Park                            4.0
Beer Garden                     3.0
Cemetery                        3.0
Coffee Shop                     3.0
Field                           3.0
Elementary School               2.0
Soccer Field                    2.0
Public Art                      2.0
Cultural Center                 2.0
Stables                         2.0
Gym                             2.0
Mexican Restaurant              2.0
Athletics & Sports              2.0
Taco Place                      2.0
Cupcake Shop                    1.0
Dessert Shop                    1.0
Dance Studio                    1.0
Dentist's Office                1.0
College Library                 1.0
Event Space                     1.0
College Rec Center              1.0
University                      1.0
College Classroom               1.0
College Auditorium              1.0
College Academic Building       1.0
Food & Drink Shop           

Here are the results for the remaining 6 clusters:

In [55]:
mcv1 = cluster1.groupby('1st Most Common Venue').count()
mcv2 = cluster1.groupby('2nd Most Common Venue').count()
mcv3 = cluster1.groupby('3rd Most Common Venue').count()
mcv4 = cluster1.groupby('4th Most Common Venue').count()
mcv5 = cluster1.groupby('5th Most Common Venue').count()
mcv6 = cluster1.groupby('6th Most Common Venue').count()
mcv7 = cluster1.groupby('7th Most Common Venue').count()
mcv8 = cluster1.groupby('8th Most Common Venue').count()
mcv9 = cluster1.groupby('9th Most Common Venue').count()
mcv10 = cluster1.groupby('10th Most Common Venue').count()
tmcv = mcv1.add(mcv2.add(mcv3.add(mcv4.add(mcv5.add(mcv6.add(mcv7.add(mcv8.add(mcv9.add(mcv10, fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0)
print('{} records in cluster 1, 10 top venues = {} entries'.format(str(cluster1.shape[0]),str(int(tmcv['PlaceName'].sum()))))
tmcv['PlaceName'].sort_values(ascending=False)

6 records in cluster 1, 10 top venues = 60 entries


Mexican Restaurant         6.0
Clothing Store             4.0
Country Dance Club         4.0
Argentinian Restaurant     3.0
Cupcake Shop               3.0
Convenience Store          3.0
Soccer Field               3.0
Dentist's Office           3.0
Comfort Food Restaurant    2.0
Dessert Shop               2.0
Market                     2.0
Garden                     2.0
Funeral Home               2.0
Dance Studio               2.0
College Library            1.0
College Rec Center         1.0
Business Service           1.0
Cultural Center            1.0
Brewery                    1.0
Strip Club                 1.0
Distribution Center        1.0
Farm                       1.0
Street Fair                1.0
Frozen Yogurt Shop         1.0
Housing Development        1.0
Mountain                   1.0
Pizza Place                1.0
Pool                       1.0
Rest Area                  1.0
Scenic Lookout             1.0
Shipping Store             1.0
Stables                    1.0
French R

In [56]:
mcv1 = cluster2.groupby('1st Most Common Venue').count()
mcv2 = cluster2.groupby('2nd Most Common Venue').count()
mcv3 = cluster2.groupby('3rd Most Common Venue').count()
mcv4 = cluster2.groupby('4th Most Common Venue').count()
mcv5 = cluster2.groupby('5th Most Common Venue').count()
mcv6 = cluster2.groupby('6th Most Common Venue').count()
mcv7 = cluster2.groupby('7th Most Common Venue').count()
mcv8 = cluster2.groupby('8th Most Common Venue').count()
mcv9 = cluster2.groupby('9th Most Common Venue').count()
mcv10 = cluster2.groupby('10th Most Common Venue').count()
tmcv = mcv1.add(mcv2.add(mcv3.add(mcv4.add(mcv5.add(mcv6.add(mcv7.add(mcv8.add(mcv9.add(mcv10, fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0)
print('{} records in cluster 2, 10 top venues = {} entries'.format(str(cluster2.shape[0]),str(int(tmcv['PlaceName'].sum()))))
tmcv['PlaceName'].sort_values(ascending=False)

5 records in cluster 2, 10 top venues = 50 entries


Housing Development                         4.0
Other Great Outdoors                        4.0
Movie Theater                               3.0
Rest Area                                   2.0
Field                                       2.0
High School                                 2.0
BBQ Joint                                   2.0
Bank                                        1.0
Breakfast Spot                              1.0
Church                                      1.0
Hardware Store                              1.0
Convenience Store                           1.0
Cultural Center                             1.0
Cupcake Shop                                1.0
Assisted Living                             1.0
Art Gallery                                 1.0
Dog Run                                     1.0
Festival                                    1.0
Argentinian Restaurant                      1.0
General College & University                1.0
College Rec Center                      

In [57]:
mcv1 = cluster3.groupby('1st Most Common Venue').count()
mcv2 = cluster3.groupby('2nd Most Common Venue').count()
mcv3 = cluster3.groupby('3rd Most Common Venue').count()
mcv4 = cluster3.groupby('4th Most Common Venue').count()
mcv5 = cluster3.groupby('5th Most Common Venue').count()
mcv6 = cluster3.groupby('6th Most Common Venue').count()
mcv7 = cluster3.groupby('7th Most Common Venue').count()
mcv8 = cluster3.groupby('8th Most Common Venue').count()
mcv9 = cluster3.groupby('9th Most Common Venue').count()
mcv10 = cluster3.groupby('10th Most Common Venue').count()
tmcv = mcv1.add(mcv2.add(mcv3.add(mcv4.add(mcv5.add(mcv6.add(mcv7.add(mcv8.add(mcv9.add(mcv10, fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0)
print('{} records in cluster 3, 10 top venues = {} entries'.format(str(cluster3.shape[0]),str(int(tmcv['PlaceName'].sum()))))
tmcv['PlaceName'].sort_values(ascending=False)

6 records in cluster 3, 10 top venues = 60 entries


Mexican Restaurant    6.0
Auto Dealership       4.0
Pizza Place           4.0
Taco Place            4.0
Plaza                 4.0
Bank                  4.0
Café                  2.0
Soccer Stadium        2.0
Scenic Lookout        2.0
Church                2.0
Office                2.0
Internet Cafe         2.0
Dry Cleaner           2.0
Dentist's Office      2.0
Dance Studio          2.0
Campaign Office       1.0
Country Dance Club    1.0
Basketball Court      1.0
Bar                   1.0
Coffee Shop           1.0
Convenience Store     1.0
University            1.0
Cultural Center       1.0
Food & Drink Shop     1.0
Recreation Center     1.0
Restaurant            1.0
Speakeasy             1.0
Stables               1.0
Street Fair           1.0
Student Center        1.0
Ice Cream Shop        1.0
Name: PlaceName, dtype: float64

In [58]:
mcv1 = cluster4.groupby('1st Most Common Venue').count()
mcv2 = cluster4.groupby('2nd Most Common Venue').count()
mcv3 = cluster4.groupby('3rd Most Common Venue').count()
mcv4 = cluster4.groupby('4th Most Common Venue').count()
mcv5 = cluster4.groupby('5th Most Common Venue').count()
mcv6 = cluster4.groupby('6th Most Common Venue').count()
mcv7 = cluster4.groupby('7th Most Common Venue').count()
mcv8 = cluster4.groupby('8th Most Common Venue').count()
mcv9 = cluster4.groupby('9th Most Common Venue').count()
mcv10 = cluster4.groupby('10th Most Common Venue').count()
tmcv = mcv1.add(mcv2.add(mcv3.add(mcv4.add(mcv5.add(mcv6.add(mcv7.add(mcv8.add(mcv9.add(mcv10, fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0)
print('{} records in cluster 4, 10 top venues = {} entries'.format(str(cluster4.shape[0]),str(int(tmcv['PlaceName'].sum()))))
tmcv['PlaceName'].sort_values(ascending=False)

3 records in cluster 4, 10 top venues = 30 entries


Field                      3.0
Scenic Lookout             3.0
Convenience Store          3.0
School                     2.0
Mexican Restaurant         2.0
Italian Restaurant         2.0
Taco Place                 2.0
Country Dance Club         1.0
Campground                 1.0
Clothing Store             1.0
College Rec Center         1.0
Comfort Food Restaurant    1.0
Toll Booth                 1.0
Distribution Center        1.0
Electronics Store          1.0
French Restaurant          1.0
Hardware Store             1.0
Mountain                   1.0
Salon / Barbershop         1.0
Assisted Living            1.0
Name: PlaceName, dtype: float64

In [59]:
mcv1 = cluster5.groupby('1st Most Common Venue').count()
mcv2 = cluster5.groupby('2nd Most Common Venue').count()
mcv3 = cluster5.groupby('3rd Most Common Venue').count()
mcv4 = cluster5.groupby('4th Most Common Venue').count()
mcv5 = cluster5.groupby('5th Most Common Venue').count()
mcv6 = cluster5.groupby('6th Most Common Venue').count()
mcv7 = cluster5.groupby('7th Most Common Venue').count()
mcv8 = cluster5.groupby('8th Most Common Venue').count()
mcv9 = cluster5.groupby('9th Most Common Venue').count()
mcv10 = cluster5.groupby('10th Most Common Venue').count()
tmcv = mcv1.add(mcv2.add(mcv3.add(mcv4.add(mcv5.add(mcv6.add(mcv7.add(mcv8.add(mcv9.add(mcv10, fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0)
print('{} records in cluster 5, 10 top venues = {} entries'.format(str(cluster5.shape[0]),str(int(tmcv['PlaceName'].sum()))))
tmcv['PlaceName'].sort_values(ascending=False)

2 records in cluster 5, 10 top venues = 20 entries


BBQ Joint                  2.0
General Entertainment      2.0
Diner                      2.0
Taco Place                 1.0
Distribution Center        1.0
Bar                        1.0
Cafeteria                  1.0
College Classroom          1.0
Comfort Food Restaurant    1.0
Event Space                1.0
Soccer Field               1.0
Field                      1.0
Food Truck                 1.0
Garden                     1.0
Nail Salon                 1.0
Park                       1.0
Auto Garage                1.0
Name: PlaceName, dtype: float64

In [60]:
mcv1 = cluster6.groupby('1st Most Common Venue').count()
mcv2 = cluster6.groupby('2nd Most Common Venue').count()
mcv3 = cluster6.groupby('3rd Most Common Venue').count()
mcv4 = cluster6.groupby('4th Most Common Venue').count()
mcv5 = cluster6.groupby('5th Most Common Venue').count()
mcv6 = cluster6.groupby('6th Most Common Venue').count()
mcv7 = cluster6.groupby('7th Most Common Venue').count()
mcv8 = cluster6.groupby('8th Most Common Venue').count()
mcv9 = cluster6.groupby('9th Most Common Venue').count()
mcv10 = cluster6.groupby('10th Most Common Venue').count()
tmcv = mcv1.add(mcv2.add(mcv3.add(mcv4.add(mcv5.add(mcv6.add(mcv7.add(mcv8.add(mcv9.add(mcv10, fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0), fill_value=0)
print('{} records in cluster 5, 10 top venues = {} entries'.format(str(cluster6.shape[0]),str(int(tmcv['PlaceName'].sum()))))
tmcv['PlaceName'].sort_values(ascending=False)

2 records in cluster 5, 10 top venues = 20 entries


Stationery Store           2.0
Outdoors & Recreation      2.0
Food Stand                 2.0
Festival                   2.0
University                 1.0
Taco Place                 1.0
Stables                    1.0
Pool                       1.0
Pharmacy                   1.0
Outdoor Sculpture          1.0
Miscellaneous Shop         1.0
Health & Beauty Service    1.0
Food Truck                 1.0
Diner                      1.0
Dessert Shop               1.0
Cupcake Shop               1.0
Name: PlaceName, dtype: float64

The results of this project have shown how Data Science algorithms and tools can help to find hidden information in large datasets. Is really important to notice that with the generated information a lot of different analysis can be done, defining clear objectives a big bunch of questions will be answered with certainty and confidence. Maybe some of the results were expected like the one that shows that a _Mexican Restaurant or a Taco Place_ are part of the top venue categories because they appear in at least 5 of the 7 clusters, but we are mexicans, right? There are some other results that were difficult to anticipate, e.g. `cluster3` is a good place to start an _Auto Dealership_ business. Besides, opening a _Sporting Goods Shop_ nearby a _Soccer Field_ venue would be a good idea to start a new business. A _Soccer Field_ venues appears in 4 of the 7 clusters and there is no trending _Sporting Goods Shop_ close those. Long story short, the available information can be used to make better decisions. 

<font size="1">[Go to Detailed results & discussion](#DetRe)</font>

<font size="1">[Go to Table of contents](#TaOCo)</font>

### <a name="Concl">Conclusions</a>

I hope you agree with me that the analysis done in this project can be applied broadly if and only if the essential data is present for the city or group of cities where the model wants to be applied. Adding more available data to the model in the future, like real estate costs or criminal rate, will make it superior. On the other hand, the developed code is easy to follow to make a similar analysis for any part of the world.

> _In the business people with expertise, experience and evidence will make more profitable decisions than people with instinct, intuition and imagination._ <font size="1">Amit Kalantri https://www.goodreads.com/quotes/tag/data-science</font>

<font size="1">[Go to Table of contents](#TaOCo)</font>

### <a name="Refer"></a>References

- [Notebook image](https://www.garybarker.co.uk)
- [Volatile economies article](https://qz.com/1550062/the-most-and-least-volatile-economies-of-the-21st-century/)
- [Geolocation Mexico Postal Codes](http://download.geonames.org/export/zip/)
- [Foursquare endpoints](https://developer.foursquare.com/docs/api/endpoints)
- [Data Science framework](https://labs.cognitiveclass.ai)
- [k-means clustering](https://en.wikipedia.org/wiki/K-means_clustering)
- [Why K-means centroids contain NAN?](https://github.com/src-d/kmcuda/issues/61)

<font size="1">[Go to Table of contents](#TaOCo)</font>

In [69]:
cluster3

PlaceName 1st Most Common Venue 2nd Most Common Venue  \
4   San Juan, San Juan Tepenahuac            Taco Place       Auto Dealership   
6           San Francisco Tecoxpa                Office    Mexican Restaurant   
7         San Salvador Cuauhtenco    Mexican Restaurant               Stables   
8          San Jerónimo Miacatlán         Internet Cafe                  Bank   
17                       Ocotitla    Mexican Restaurant                  Bank   
40           San Andrés Ahuayucan         Internet Cafe        Soccer Stadium   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4            Pizza Place     Recreation Center        Student Center   
6         Scenic Lookout           Dry Cleaner                 Plaza   
7                  Plaza            Taco Place                Church   
8     Mexican Restaurant           Pizza Place                 Plaza   
17        Soccer Stadium            Taco Place       Auto Dealership   
40    Mexican Restaurant           Coffee Shop            Taco Place   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4     Mexican Restaurant           Street Fair                Church   
6            Pizza Place                  Bank                  Café   
7                    Bar            Restaurant             Speakeasy   
8                   Café        Scenic Lookout           Dry Cleaner   
17      Basketball Court                 Plaza           Pizza Place   
40       Campaign Office        Ice Cream Shop     Convenience Store   

   9th Most Common Venue 10th Most Common Venue  
4                   Bank           Dance Studio  
6        Auto Dealership      Food & Drink Shop  
7       Dentist's Office           Dance Studio  
8                 Office        Auto Dealership  
17            University       Dentist's Office  
40    Country Dance Club        Cultural Center